lib/docker.js

```diff
@@ -0,0 +1,454 @@
+                }).then(function(logs) {
+                    if(!logs && !opts.need_build) {
+                        // no updates have happened, nothing to do here any more but clean up
+                        // go back to the root dir
+                        process.chdir(opts.dir);
+                        // and get back to master
+                        return promised_git(['checkout', 'master'])
+                        .then(function() {
+                            console.log('The deploy repository is up to date already, exiting.');
+                            process.exit(0);
+                        });
+                    } else if(logs) {
+                        logs += "\n";
+                    } else if(!logs) {
+                        logs = '';
+                    }
+                    opts.commit_msg += "List of changes:\n" + logs;
+                    return promised_git(['checkout', 'origin/master']);
+                }).then(function() {
+                    // go back to the root dir
+                    process.chdir(opts.dir);
+                    // add the submodule changes
+                    return promised_git(['add', opts.submodule]);
+                });
+            });
+        } else {
+            // no submodule, need to add it
+            opts.submodule = 'src';
+            opts.need_build = true;
+            opts.commit_msg = 'Initial import of ' + opts.name;
+            return promised_git(['submodule', 'add', 'https://gerrit.wikimedia.org/r/mediawiki/services/' + opts.name, opts.submodule]);
+        }
+    }).then(function() {
+        // make sure the package.json symlink is in place
+        return fs.symlinkAsync(opts.submodule + '/package.json', 'package.json')
+        .catch(function() {}).then(function() {
+            return promised_git(['add', 'package.json']);
+        });
+    }).then(function() {
+        if(!opts.need_build) {
+            return;
+        }
+        // update the commit message
+        if(!/^initial/i.test(opts.commit_msg)) {
+            opts.commit_msg += "xxxxxxx Update node module dependencies\n";
+        }
+        // a rebuild is needed, start by removing the existing modules
+        return promised_git(['rm', '-r', 'node_modules'])
+        .then(function() {
+            return promised_spawn(['rm', '-rf', 'node_modules'], true);
+        }).then(function() {
+            // start the container which builds the modules
+            return start_container(['-v', opts.dir + ':/opt/service', '-w', '/opt/service']);
+        }).then(function() {
+            // remove unnecessary files
+            return promised_spawn([
+                'find',
+                'node_modules/',
+                '-regextype',
+                'posix-egrep',
+                '-iregex',
+                "(.*\\.git.*|.*\\.md|.*\\.txt|.*readme|.*licence)",
```

Comment:
Afaik the only things we really *need* to remove are `.gitignore` files, as they'll mess with the checkin. Removing the other files you match here is *probably* fine, but I'm not sure that it's worth it.